### PR TITLE
fix: Sort breadcrumbs on timestamp before rendering

### PR DESCRIPTION
**Pull request by kamilogorek**
Reference: https://forum.sentry.io/t/bad-sorting-of-manual-breadcrumbs/3630

At first, I fixed it in the JS SDK https://github.com/getsentry/raven-js/pull/1308, however, after rethinking it, I believe we should do this here, so it's fixed for every SDK instead.

It's out of order, because `timestamp` is one of the parameters that can be passed to `captureBreadcrumb`, yet we push them at the end of the breadcrumbs array as they happen.

### CODE REVIEW

**Comment by MaxBittker** [comment]
Awesome, not sure what percentage of events will be effected, but this seems much more correct.

How well does this deal with undefined timestamps - is there a risk that any of the major sdks are sending differently formed data for this key here? or do we normalize them on the server? 

the percy snapshot breadcrumbs have been re-ordered - the exceptions seem to all get moved to the top. is this just a problem with that hardcoded mock data we should fix?

**Comment by kamilogorek** [comment]
I somehow don't have access to Percy to verify this. I checked components that render breadcrumbs and they don't handle missing data, so I assumed they have to be unified at one point if we already allowed that. We might have to double-check it.

@dcramer can you give me Percy access?

File: src/sentry/static/sentry/app/components/events/interfaces/breadcrumbs.jsx [hunk 1]
```diff
@@ -68,6 +68,7 @@ class BreadcrumbsInterface extends React.Component {
     // reverse array to get consistent idx between collapsed/expanded state
     // (indexes begin and increment from last breadcrumb)
     return crumbs
+      .sort((a, b) => a.timestamp > b.timestamp)
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/components/events/interfaces/breadcrumbs.jsx]: The results in Percy suggest this isn't sorting correctly.

`.sort` should return -1, 0, or 1. Is it possible this is the cause?

**Review by benvinegar — CHANGES_REQUESTED** [review]
`.sort` needs to return -1, 0, or 1.

**Comment by kamilogorek** [comment]
@MaxBittker about undefined timestamps, it cannot happen, we have a check for this - https://github.com/getsentry/sentry/blob/96bc1c63df5ec73fe12c136ada11561bf52f1ec9/src/sentry/interfaces/breadcrumbs.py#L66-L75

@benvinegar `true/false` implicitly coerces to `1/0` and we care only about whether we return `1` or not in this case. 

The issue, however, is much more complicated than this and took me a while to find the main cause. The first crumb, which is the main exception is simply created not on the same day.

Issue flow:
- we render breadcrumbs based on provided data https://github.com/getsentry/sentry/blob/aedcab8f45161ae9d607da273f1fd44db17ba23e/src/sentry/static/sentry/app/components/events/interfaces/breadcrumbs.jsx#L216
- data is extracted from the event and only then we add something called Virtual Crumb, which is created from the event itself https://github.com/getsentry/sentry/blob/aedcab8f45161ae9d607da273f1fd44db17ba23e/src/sentry/static/sentry/app/components/events/interfaces/breadcrumbs.jsx#L195-L199
- this crumb's date is based on `dateCreated` attribute https://github.com/getsentry/sentry/blob/aedcab8f45161ae9d607da273f1fd44db17ba23e/src/sentry/static/sentry/app/components/events/interfaces/breadcrumbs.jsx#L121-L126
- our acceptance tests are creating sample event with a fixed timestamp https://github.com/getsentry/sentry/blob/master/tests/acceptance/test_issue_details.py#L23-L35
- however, `create_sample_event` implementation is overriding every breadcrumb's timestamp based on the current date, but event's date stay the same https://github.com/getsentry/sentry/blob/e112b9b51884b01b8f79629ed1a68789a31c3079/src/sentry/utils/samples.py#L176-L192

tl;dr (as I believe it works): created sample event uses fixed timestamp, where breadcrumbs are updated based on the current date (which can be seen on Percy's screenshots)

Also, how about doing this operation in the interface instead of UI? https://github.com/getsentry/sentry/blob/96bc1c63df5ec73fe12c136ada11561bf52f1ec9/src/sentry/interfaces/breadcrumbs.py#L53-L63

**Comment by mitsuhiko** [comment]
I'm still super -1 on this. What if we only sort a certain subset of breadcrumbs? JavaScript time is fundamentally nonmonotonic and local so breadcrumbs are likely to be fucked whenever the system time readjusts or we get to a timezone change.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]
I really do not want to do this. We could offer it as an option (toggle between order of submission / timestamp) optionally.

**Comment by kamilogorek** [comment]
I'm totally fine keeping it in `dataCallback` of JS SDK as well and won't hesitate to close this pr.

**Comment by benvinegar** [comment]
> @benvinegar true/false implicitly coerces to 1/0 and we care only about whether we return 1 or not in this case.

It's not that it coerces to 1/0. My understanding is that because depending on the sorting algortihm, 1 and 0 is not enough to produce stable results depending on the browser's internal sorting algorithm, which is not standardized.

More here: https://stackoverflow.com/a/234777

For example, this occurs in Chrome right now:

```javascript
[5, 8, 7, 1, 2, 3, 4, 6, 9, 10, 11, 12, 13].sort(function(a, b) {
    return a > b;
});
// => [4, 5, 3, 1, 2, 6, 7, 8, 9, 10, 11, 12, 13]
```

**Comment by kamilogorek** [comment]
Not relevant anymore due to v4 SDK.
